Feature/haeun step2

pear96 opened the pull request:
[FolderMetaData] - 폴더 메타 데이터
- 폴더 pk
- 폴더 이름 / 사용자 이름
- 폴더 경로 (폴더 이름을 포함한 절대 경로)
- 부모 폴더 pk (findAllByParentFolderId를 위해 인덱스 설정)
- 폴더 이름 변경을 위해 폴더 이름과 경로를 바꿔주는 메서드 구현

[FolderRepository]
- 폴더 아이디를 기반으로 폴더의 경로만 찾아오는 메서드는 파일의 경로를 찾을 때 필요합니다. 파일에는 더 이상 절대 경로가 아닌 서버 파일명만 저장하고, 해당 파일을 실제로 읽을 때에는 `해당 파일이 존재하는 폴더의 경로 + 파일명`으로 읽습니다. 폴더 경로 외의 다른 정보는 불필요하므로 쿼리를 작성했습니다.
- (사용자 이름 + 폴더 이름 + 부모 폴더 pk) 로 폴더 데이터를 조회하는 이유는, 폴더 중복 생성을 막기 위함입니다.

[FolderService]
- 조회, 생성, 수정 모두 **부모 폴더를 확인한** 뒤 로직을 수행합니다.
- 부모 폴더 확인을 위해 모든 폴더 응답에는 폴더의 기본키를 포함했습니다. (다른 정보로 바꿔야할까요..)
- 폴더 데이터 조회
  폴더의 이름, 사용자 이름, 하위 폴더 목록(폴더 이름, 사용자 이름), 하위 파일 목록을 반환합니다.
- 폴더 생성
  1. 부모 확인, 중복 확인
  2. 부모 경로 + 폴더 이름 = 폴더 경로
  3. 폴더 생성
  4. 폴더 데이터 저장
- 폴더 이름 수정
  - 가장 고민이 많았던 부분입니다. 결과적으로는 하위 폴더의 경로 변경을 위해 **BFS**를 선택했습니다. 그러나 하위 폴더의 기본키를 부모 폴더 키로 가진 폴더들을 계속 확인한다는 점에서 DB에 매우 많은 요청이 갈 수 있기 때문에 불안합니다...
  - 만약 폴더에 절대 경로를 저장하지 않는다면 폴더 이름만 바꾸면 되겠지만, 그럼 폴더 경로를 조회하기 위해 매번 루트 폴더를 DFS로 탐색해야 한다고 생각했습니다. 조회가 무거운 것 보다는 수정이 무거운게 나을 것 같아서 폴더 명을 수정하면 하위 폴더들의 경로를 수정하도록 구현했습니다.
  - BFS 탐색에 사용하려고 FolderInfo라는 내부 클래스를 만들었습니다.

[ExceptionHandlingController]
`@Valid` 어노테이션으로 인해 발생하는 MethodArgumentNotValidException 처리를 추가했습니다.
사실은 ServiceException과 그 외는 전부 500으로만 처리하고 싶어서 AOP를 적용해 MethodArgumentNotValidException와 ValidationException가 발생하면 ServiceException을 던지게 하려고 했습니다. 그러나 `@AfterThrow`에서 어떤 예외인지 판별할 수 없어서 결국 메서드 내에서 분리해야했고 이는 의미가 없다 판단해 ExceptionHandlingController에 분리해서 작성했습니다. UnchekedException이지만 500이 아닌 400 응답코드를 반환해야하는데 고민입니다.

그외 파일 로직에 폴더 아이디를 추가하고, 테스트 코드를 작성했습니다.

pear96 committed:
chore: yml profile 설정

dev 상태로 분리해주었습니다.

pear96 committed:
refactor: 예외 처리

@Valid 어노테이션으로 인해 발생하는 MethodArgumentNotValidException 처리를 추가했습니다.
사실은 ServiceException과 그 외만 처리하고 싶어서 AOP를 적용해 MethodArgumentNotValidException와 ValidationException가 발생하면 ServiceException을 던지게 하려고 했습니다. 그러나 @AfterThrow에서 어떤 예외인지 판별할 수 없어서 결국 메서드 내에서 분리해야했고 이는 의미가 없다 판단해 ExceptionHandlingController에 분리해서 작성했습니다. UnchekedException이지만 500이 아닌 400 응답코드를 반환해야하는데 고민입니다.

pear96 committed:
feat: 회원가입

`서버_물리_저장소_루트` 위치에서 각 사용자별 폴더를 생성하기 위해 회원가입만 하는 기능을 추가했습니다. (사용자명 중복 불가)
따라서 서버의 DB에는 `서버_물리_저장소_루트` 가 기본키를 0으로 해서 추가되고, 이후 사용자 이름으로 만든 폴더가 생기며 부모 폴더의 pk를 0으로 설정합니다.

pear96 committed:
feat: 폴더 기능

[FolderMetaData] - 폴더 메타 데이터
- 폴더 pk
- 폴더 이름 / 사용자 이름
- 폴더 경로 (폴더 이름을 포함한 절대 경로)
- 부모 폴더 pk (findAllByParentFolderId를 위해 인덱스 설정)
- 폴더 이름 변경을 위해 폴더 이름과 경로를 바꿔주는 메서드 구현

[FolderRepository]
- 폴더 아이디를 기반으로 폴더의 경로만 찾아오는 메서드는 파일의 경로를 찾을 때 필요합니다. 파일에는 더 이상 절대 경로가 아닌 서버 파일명만 저장하고, 해당 파일을 실제로 읽을 때에는 `해당 파일이 존재하는 폴더의 경로 + 파일명`으로 읽습니다. 폴더 경로 외의 다른 정보는 불필요하므로 쿼리를 작성했습니다.
- (사용자 이름 + 폴더 이름 + 부모 폴더 pk) 로 폴더 데이터를 조회하는 이유는, 폴더 중복 생성을 막기 위함입니다.

[FolderService]
- 조회, 생성, 수정 모두 부모 폴더를 확인한 뒤 로직을 수행합니다.
- 부모 폴더 확인을 위해 모든 폴더 응답에는 폴더의 기본키를 포함했습니다. (다른 정보로 바꿔야할까요..)
- 폴더 데이터 조회
  폴더의 이름, 사용자 이름, 하위 폴더 목록(폴더 이름, 사용자 이름), 하위 파일 목록을 반환합니다.
- 폴더 생성
  1. 부모 확인, 중복 확인
  2. 부모 경로 + 폴더 이름 = 폴더 경로
  3. 폴더 생성
  4. 폴더 데이터 저장
- 폴더 이름 수정
  - 가장 고민이 많았던 부분입니다. 결과적으로는 하위 폴더의 경로 변경을 위해 BFS를 선택했습니다. 그러나 하위 폴더의 기본키를 부모 폴더 키로 가진 폴더들을 계속 확인한다는 점에서 DB에 매우 많은 요청이 갈 수 있기 때문에 불안합니다...
  - 만약 폴더에 절대 경로를 저장하지 않는다면 폴더 이름만 바꾸면 되겠지만, 그럼 폴더 경로를 조회하기 위해 매번 루트 폴더를 DFS로 탐색해야 한다고 생각했습니다. 조회가 무거운 것 보다는 수정이 무거운게 나을 것 같아서 폴더 명을 수정하면 하위 폴더들의 경로를 수정하도록 구현했습니다.
  - BFS 탐색에 사용하려고 FolderInfo라는 내부 클래스를 만들었습니다.

pear96 committed:
refactor: 파일 관련 로직에 폴더 id 추가

[FileMetaData]
- 파일이 위치한 폴더의 기본키인 folderId를 컬럼에 추가하고,인덱스를 적용했습니다. 폴더 정보 조회시 findAllByFolderId를 수행 해야하기 때문입니다.

[FileController]
- 기존에 파일 조회시 @requestpart 로 파라미터를 받았는데, DTO로 변환했습니다.
- 파일 다운로드시 Resource만 반환했더니 응답이 인코딩이 되지 않은 상태로 와서 ResponseEntity에 MediaType을 설정해 반환했습니다.

[FileService]
- 폴더가 실제로 존재하는지 확인하는 로직을 추가했습니다.

그외 응답 DTO를 생성했습니다.

pear96 committed:
test: 폴더 테스트 코드 추가

- 테스트 데이터 공유로 인해 테스트 실패하는 경우가 생겨서 기본적인 값 외에는 테스트 마다 데이터를 생성했습니다.

[FolderServiceTest]
- 폴더 조회 성공
  - 하위 폴더 조회에 성공하는지 확인하기 위해 변수가 많습니다...
  - `ReflectionTestUtils`를 활용해 폴더 메타 데이터에 pk를 설정해 분리하려고 했습니다.
- 폴더 이름 수정 성공
  - `ReflectionTestUtils`가 가징 필요했던 부분입니다. 폴더 pk를 부모 pk로 가진 하위 폴더들을 찾아야 하는데 부모 폴더 메타데이터와 자식 폴더 메타데이터의 pk가 둘다 0이라 while문이 종료되지 않는 문제가 있었습니다.

[FolderControllerTest]
- 폴더 조회 / 생성 / 폴더 이름 수정

[FileUtilTest]
- 폴더 생성 / 폴더 생성 실패
- 폴더 이름 변경 / 폴더 이름 변경 실패

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/dto/request/UploadFileReq.java:
```diff
@@ -0,0 +1,13 @@
+public record UploadFileReq(
+	@NotNull(message = "file doesn't exist") MultipartFile file,
+	@NotBlank(message = "user name is blank") String userName,
+	@NotNull(message = "folder id is null") Long folderId
```
long 으로 하는 건 어떤 가요

pear96 replied:
헐 그러게요 계속 pk는 long으로 하고 있었는데 이걸 이렇게.. 잡아주셔서 감사합니다.

kochungcheon commented on src/test/java/com/c4cometrue/mystorage/TestParameter.java:
```diff
@@ -5,9 +5,13 @@
+import com.c4cometrue.mystorage.entity.FolderMetaData;
 
 public class TestParameter {
     public static final MultipartFile mockMultipartFile = mock(MultipartFile.class);
```
상수는 대문자로 표기하는 거 어떤가요

pear96 replied:
아 그러게요! 좋은 질문? 지적? 감사합니다.
![수정됨_72934_74447_5624](https://github.com/C4-ComeTrue/my-storage/assets/80267948/92cad910-cf82-4ffe-a589-4e85bb0d1703)

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/controller/FolderController.java:
```diff
@@ -0,0 +1,48 @@
+
+	private final FolderService folderService;
+
+	@GetMapping()
```
() 제거 하는 거 어떤 가요

pear96 replied:
ㅋㅋㅋㅋㅋ 아주 좋습니다! (하 왜 자꾸 이러지..)

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/entity/FileMetaData.java:
```diff
@@ -13,32 +15,41 @@
-    }
+	@Id
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
+	private long fileId;
```
id 는 Long 으로 하는 거 어떤가요
https://docs.jboss.org/hibernate/orm/5.3/userguide/html_single/Hibernate_User_Guide.html#entity-pojo-identifier

pear96 replied:
음 이해했습니다.
만약 long으로 하면 기본값이 0으로 초기화 되는데, 여러 엔티티를 동시에 저장하다 문제가 생겨버려서 모두 같은 값을 가지게 되어버리면, 문제 지점을 파악하는게 Null인 것 보다 어렵다는 단점이 있군요. 기본 키를 외부에서 주입 받을 수도 있고...
감사합니다!

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/repository/FolderRepository.java:
```diff
@@ -0,0 +1,23 @@
+
+import com.c4cometrue.mystorage.entity.FolderMetaData;
+
+@Repository
```
이 부분은 없어도 될 거 같아요!

pear96 replied:
JpaRepository를 상속받기 때문이군요

pear96 replied:
from 청천

https://www.inflearn.com/questions/227764/extends-jparepository-lt-gt-vs-repository-%EC%A7%88%EB%AC%B8
https://velog.io/@kshired/Spring-%EC%99%9C-JPARepository%EB%8A%94-Repository%EA%B0%80-%ED%95%84%EC%9A%94-%EC%97%86%EC%9D%84%EA%B9%8C-deep-dive-%ED%95%B4%EB%B3%B4%EA%B8%B0

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -8,98 +8,128 @@
+			file.getOriginalFilename(), uploadFileReq.userName(), uploadFileReq.folderId()).isPresent()) {
+			throw ErrorCd.DUPLICATE_FILE.serviceException();
+		}
+		String fileStorageName = UUID.randomUUID() + file.getOriginalFilename();
```
UUID생성은 다른 메서드나 File 엔티티 안에서 처리 하는 거 어떤 가요

pear96 replied:
흠.. 메서드 명으로 무엇을 하는지만 알고 실제 작동 코드는 볼 필요가 없어지기 때문인가요??

kochungcheon replied:
그렇습니다 !

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -8,98 +8,128 @@
+	 * @param uploadFileReq 파일(MultiPartFile)과 사용자 이름, 폴더 아이디 포함
+	 * @return 파일의 메타 데이터
+	 */
+	@Transactional
```
트랜잭션이 오래 열릴 수도 있을 거 같아요
예를 들어 대용량 파일을 업로드할 때를 들 수 있을 거 같아요

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -8,98 +8,128 @@
+		Path filePath = folderPath.resolve(fileMetaData.getFileStorageName());
+		// 파일 데이터 로드
+		Resource file = resourceLoader.getResource("file:" + filePath.toString());
+		if (!file.exists()) {
```
별도로 분리하는 건 어떤 가요

pear96 replied:
적극 반영하겠습니다.

VSFe commented on src/main/java/com/c4cometrue/mystorage/dto/request/CreateFolderReq.java:
```diff
@@ -0,0 +1,13 @@
+import jakarta.validation.constraints.NotNull;
+
+/**
+ * DTO for {@link com.c4cometrue.mystorage.entity.FolderMetaData}
```
👍
근데 그냥 `@See com.c4cometrue.mystorage.entity.FolderMetaData` 라고 쓰는게 나을수도?

pear96 replied:
사실 intellij에서 DTO로 만들어서 자동 생성된 주석이었는데.. 
`@link`는 인라인으로, `@see` 는 독립적인 블록으로 생성된다는 차이점이 있군요
DTO들은 다 Entity 기반으로 생성한거고 특별한 설명이 필요 없으니 `@see`만 써도 되겠군요.
https://stackoverflow.com/questions/10097199/javadoc-see-or-link

VSFe commented on src/main/java/com/c4cometrue/mystorage/dto/request/GetFolderReq.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * DTO for {@link com.c4cometrue.mystorage.entity.FolderMetaData}
+ */
+public record GetFolderReq(
```
이건 진짜 그냥 개인적인 스타일이긴 한데 ㅋㅋㅋ

```java
public record GetFolderReq(
	@NotNull long folderId,
	@NotBlank(message = "folder name is blank") String folderName,
	@NotBlank(message = "user name is blank") String userName,
	@NotNull long parentFolderId
) {
}
```
처럼 쓰는게 좀 더 낫더라고요.

pear96 replied:
호오 그렇군요..!

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/UserService.java:
```diff
@@ -0,0 +1,55 @@
+			throw ErrorCd.DUPLICATE_USER.serviceException();
+		}
+		// 유저 root 폴더 생성
+		Path userFolderPath = storagePathService.createBasicFolderPath(signUpReq.userName());
```
👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍

다른 그룹에서도 이거 시켰을때 죄다 rootFolderId = -1로 설정하는데, 이러면 PK 설정 관점에서 정말 답이 없어져서 마음이 아팠었습니다...

pear96 replied:
와우 무한 붐업을 받다니 행복하군요 ^__^

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/service/FolderService.java:
```diff
@@ -0,0 +1,179 @@
+	}
+
+	@Transactional
+	public void updateFolderName(UpdateFolderNameReq updateFolderNameReq) {
```
해당 메서드가 담당하는 역할이 많은 거 같아요 분할 하는 건 어떤 가요

pear96 replied:
적극 반영해보겠습니다.

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/UserService.java:
```diff
@@ -0,0 +1,55 @@
+		FileUtil.createFolder(userFolderPath);
+
+		// 유저 생성
+		UserData user = UserData.builder()
```
요렇게 짧게 갈거면, 사실 builder는 너무 투머치 하죠?

pear96 replied:
음 맞는 말입니다... 인자가 1개 뿐이니... 생성자로 생성하는게 가독성에도 효율면에서도 좋을 것 같아요

VSFe commented on src/main/java/com/c4cometrue/mystorage/util/FileUtil.java:
```diff
@@ -3,46 +3,68 @@
+		}
+	}
+
+	public static void createFolder(Path folderPath) {
```
기왕 javadoc 다는 겸, 나머지 메서드에도 달아주세요.

pear96 replied:
🤓 네 적극 반영하겠습니다

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/service/FolderService.java:
```diff
@@ -0,0 +1,179 @@
+	}
+
+	@Transactional
+	public void updateFolderName(UpdateFolderNameReq updateFolderNameReq) {
```
루트 폴더나 정말 많은 하위 폴더나 파일을 가진 폴더의 이름을 바꾸게 되면 트랜잭션이 오래 열리게 될 거 같아요 그리고 다 바꾸는 데 발생하는 비용도 상당할 거 같아요

pear96 replied:
저도 그 부분이 걱정스러운데.. 한 값만 바꾸려면 아예 폴더 경로를 DB에 저장하지 말고 폴더를 조회할 때마다 루트까지 타고가야 할 것 같아서요 =.= 음........

VSFe replied:
적절하게 캐싱을 쓰면 되겠죠?

pear96 replied:
폴더를 DB에서만 저장하고 물리적 폴더를 생성하지 않는 것으로 바꿨습니다.
그래서 해당 폴더의 이름만 바꾸면 되기 때문에 하위 폴더는 수정할 일이 없습니다.

kochungcheon reviewed (commented):
고생하셨습니다! 특히 개별 유저 마다 폴더 만들어 준다는 전략 인상 깊게 봤습니다 많이 배우고 갑니다!

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/entity/UserData.java:
```diff
@@ -0,0 +1,26 @@
+	@NotBlank(message = "user name is blank")
+	private String userName;
+
+	@Builder
```
빌더 패턴과 정적 팩토리 패턴 이 둘을 언제 사용하면 좋을 지 항상 고민이 들던데, 하은님은 어떤 기준으로 이 둘을 사용하시나요

pear96 replied:
웬만하면 Builder를 우선으로 썼는데, 필드 수가 적어서 별로 헷갈리지 않는다면 정적 패토리 메서드가 가독성이 더 좋은 것 같긴합니다.

pear96 commented on src/main/java/com/c4cometrue/mystorage/service/FolderService.java:
```diff
@@ -0,0 +1,179 @@
+	}
+
+	private List<FileMetaDataRes> getFiles(FolderMetaData folder) {
+		List<FileMetaDataRes> fileResult = new LinkedList<>();
```
fileResult 는 쿼리로 받아온 결과를 삽입만 하고 반환하는거라 `LinkedList`를 사용했습니다.. for문 순회도 fileList에서 하는거라.. 혹시 제가 이해가 덜 된걸까요

pear96 commented on src/main/java/com/c4cometrue/mystorage/exception/ExceptionHandlingController.java:
```diff
@@ -1,51 +1,74 @@
-        // 에러 메세지 + 시간
+    @ExceptionHandler({MethodArgumentNotValidException.class})
+    public ResponseEntity<ApiExceptionRes> handleValidException(MethodArgumentNotValidException exception) {
+        String errorMessage = exception.getBindingResult().getFieldErrors().stream()
```
ㅜㅜ 이거 테스트 코드 어케 짜면 좋을까요...

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FolderService.java:
```diff
@@ -0,0 +1,179 @@
+			userName, folderName, parentFolderId);
+	}
+
+	private FolderMetaData checkParentFolder(long parentFolderId) {
```
`check` 라는 네이밍은 존재 여부를 판단하는 (즉, true, false를 반환하는) 의미가 좀 강한 것 같아요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FolderService.java:
```diff
@@ -0,0 +1,179 @@
+		);
+	}
+
+	private static class FolderInfo {
```
record

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FolderService.java:
```diff
@@ -0,0 +1,179 @@
+
+	public FolderOverviewRes getFolderData(GetFolderReq getFolderReq) {
+		// parentFolder가 없다면? -> Error (잘못된 요청)
+		checkParentFolder(getFolderReq.parentFolderId());
```
사실 이 논리면 상위 폴더 전부 다 조회해야겠죠?

pear96 replied:
하하하하하..... 맞아요... 좀 더 고민해보겠습니다.

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FolderService.java:
```diff
@@ -0,0 +1,179 @@
+		Path folderPath = storagePathService.createFolderPath(parentFolder.getFolderPath(),
+			folderName);
+		// 물리적 폴더 생성
+		FileUtil.createFolder(folderPath);
```
폴더를 물리적으로 만들 필요가 있을까요?
막말로 모든 파일을 전부 하나의 물리적 폴더에 때려 박아도 상관 없지 않을까요?
(폴더 내에 파일이 엄청나게 많은 상황에서 폴더 이동 과업을 수행하게 되면, Disk I/O가 미친듯이 발생하겠죠.)

pear96 replied:
으아 그러게요 파일 명을 물리적 서버에 저장한 이름과 DB에 원본 이름, 저장소 이름 둘 다 저장하는건 생각했었는데 왜 꼭 파일 트리를 맞춰야한다고 생각했을까요... 좋은 생각이네요

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FolderService.java:
```diff
@@ -0,0 +1,179 @@
+		folderIdList.add(init);
+
+		// BFS
+		while (!folderIdList.isEmpty()) {
```
별도의 함수로 분리하고 재귀함수 및 Stream을 사용하면, 훨씬 깔끔한 코드 작성이 가능합니다.
자세한건 나중에...

VSFe commented on src/main/java/com/c4cometrue/mystorage/repository/FileRepository.java:
```diff
@@ -9,7 +10,9 @@
 @Repository
 public interface FileRepository extends JpaRepository<FileMetaData, Long> {
-    Optional<FileMetaData> findByFileNameAndUsername(String filename, String username);
+	Optional<FileMetaData> findByFileNameAndUserNameAndFolderId(String filename, String username, long folderId);
```
인덱스가 필요할 것 같은데요.

pear96 replied:
지당하신 말씀입니다..
생각해보니 folderId(중복 안됨, long) -> userName(중복 안됨, String) -> fileName(중복 가능, String) 순으로 index를 생성하면 더 효율적일 것 같더라구요. 그래서 메서드와 파라미터 순서도 바꿨습니다. 💡
폴더에도 `findByUserNameAndFolderNameAndParentFolderId` 이 있어서 다시 `parentFolerId` -> `UserName` -> `FolderName`으로 탐색하도록 했습니다.

pear96 committed:
style : DTO 스타일 및 주석 수정

pear96 committed:
refactor : 서비스 레이어에 DTO 필드 하나씩 넘기기

- 컨트롤러에서는 DTO를 활용하고, 서비스에서는 각 필드를 하나씩 전달했다.
- DTO의 필드가 바뀌면 서비스도 다 바꿔야한다는 점에서 하나씩 전달하면 번거롭다는 단점이 있다.
- 그러나 서비스 안에서 일일이 DTO의 필드에 접근해야한다는 번거로움과 가독성이 떨어지는 단점이 있다.
  또한 DTO를 공유하는 경우는 적을 것 같아서 DTO의 필드가 바뀐다고 수정 사항이 아주 많지는 않을 것 같다.
=> 그래서 DTO를 서비스에 통째로 전달하는 방식에서 개별로 전달하는 것으로 바꾸었다.

pear96 committed:
refactor : 서비스 레이어에 DTO 필드 하나씩 넘기기

[GetFolderReq]
- 폴더 pk로 유일하게 구분할 수 있으니 폴더 이름과 부모 폴더 pk가 필요 없다고 판단했습니다.
- 원래는 부모 폴더가 유효한지 확인하는 과정이 있었는데...하..

[UploadFileReq]
- 폴더 id가 null인 경우는 없으므로 `Long`에서 `long`으로 변환

그 외에는 순서 변환 정도

pear96 committed:
refactor : 엔티티에 index 적용

[FileMetaData]
- 폴더 pk로 유일하게 구분할 수 있으니 폴더 이름과 부모 폴더 pk가 필요 없다고 판단했습니다.
- 원래는 부모 폴더가 유효한지 확인하는 과정이 있었는데...하...
(부모 폴더를 확인해야할까요?? 말까요?? 판단을 내릴 수가 없습니다.)
- folderId -> userName -> fileName 순으로 인덱스를 적용했습니다.
  - 우선은 folderId로 findAll 하는 메서드가 있기 때문에 folderId를 가장 우선에 두었습니다.

[FileRepository, UserRepository]
- JpaRepository를 상속받기 때문에 @Repository를 등록해주지 않아도 됩니다.

[FolderMetaData]
- idx를 folderName -> parentFolderId -> userName 순으로 적용했습니다.
- 원래는 parentId -> userName -> folderName이었는데, 우선 부모 폴더를 찾는 로직을 제거해서 parentId 인덱스가 필요하지 않아졌습니다.
  따라서 데이터의 중복도가 낮은 순으로 변경했습니다.
- folderPath를 제거했습니다. 폴더가 실제 Disk에 생성되지 않고 DB에 개념으로만 존재하면서 필요하지 않아졌습니다.

[FolderRepository]
- 인덱스에 적용한 순서와 일치하도록 메서드 명을 작성했습니다.

[UserData]
- 기본키 타입을 `Long`으로 바꿨습니다. `long`은 null을 가질 수 없고, 값이 정해지지 않으면 0으로 할당됩니다.
- 그렇게 되면 DB에 의도와는 다르게 0으로 저장될 수 있기 때문에 차라리 Null인 경우가 문제점을 파악하기 쉽습니다.

pear96 committed:
refactor : 서비스 코드 수정

[공통]
- 컨트롤러에서 DTO 통째로 전달받던 데이터를 필드 하나씩 받도록 수정했습니다.
- 함수 주석을 전부 작성했습니다.

[FileService]
- 부모 폴더 유효성 검사를 제거했습니다.
- Exception 처리를 위해 메서드 분리를 더 자세하게 나눴습니다.
- DB 작업은 MySQL 덕분에 롤백이 가능하지만, File은 롤백이 불가능합니다.
  따라서 DB 저장에 실패하면 File 저장도 일어나지 않도록 순서를 변경했습니다.
-  `getFilePath` = 이제 폴더는 '사용자이름'에 따른 폴더만 존재하기 때문에, 모든 파일의 경로는 `root/사용자/파일저장소이름`으로 구성됩니다.
  따라서 기존에는 폴더의 정보에 따라 경로를 반환했다면, 이제는 사용자의 이름과 파일명으로 경로를 생성합니다.

[FolderService]
- 부모 폴더 유효성 검사를 제거했습니다.
- 폴더에 대한 권한을 확인합니다.
- 폴더 생성에서 폴더 데이터를 저장하고서는 다시 할당하는데, 테스트 코드 때문에 이렇게 작성했습니다.
  - Repository를 mocking했기 때문에 pk가 실제로 생성되지 않고, 단위 테스트만 진행하기 때문에 테스트 데이터베이스를 연결하기 애매하다고 판단했습니다.
  - 따라서 저장된 값을 다시 가져와서 해당 값의 folderId를 사용하도록 했습니다. (테스트 코드에서는 mocking)
- 폴더명 수정에 성공한 경우에는 어차피 클라이언트에서 갖고있는 정보이기 때문에 응답코드만 반환해도 충분하다고 판단했습니다.
  - 또한 폴더명 수정에 따른 하위 폴더의 경로를 바꿀 필요가 없어지면서 BFS 탐색을 제거했습니다.

[UserService]
- DB 데이터를 먼저 생성한 뒤 실제 폴더를 생성하도록 했습니다. 그리고 유저 폴더의 부모 폴더 pk는 0입니다.
- 실제로 pk가 0인 폴더의 데이터가 있을 필요는 없을 것 같습니다.
  - 현재 부모 폴더를 확인하는 로직이 없지만, 있다하더라도 0이면 멈출 것이기 때문입니다.

[StoragePathService]
- 사용자 이름에 기반한 경로를 반환하는 메서드만 남았으며 이름을 직관적으로 수정했습니다.

pear96 committed:
test : 테스트 코드 수정

점검하는 로직은 동일하며, 공통으로 사용하는 테스트 파라미터를 생성했습니다.
테스트 파라미터를 공통으로 사용하면 다른 테스트에 영향을 주는 경우가 있기 때문에
변하지 않는 값에 대해서만 테스트 파라미터를 생성했습니다.

pear96 commented:
### 폴더가 제대로 존재하는지(부모 폴더를 타고 올라가서 root까지 닿는지) 확인해야 할까요?

폴더가 제대로 존재하는지 확인하려면 최상위까지 타고 올라가봐야하는데, 
이걸 한다면 폴더, 파일의 모든 로직마다 수행해야합니다.

[확인한다.]
- 요청이 많으면 병목 생기기 딱 좋음. 시간이 매우 많이 소요될 수 있을 것 같습니다..
- 스스로를 부모 폴더로 보면 안되기 때문에 이제까지 지나온 폴더들이랑 겹치지 않는지도 확인해야합니다. (HashSet..?)
- 아니면 확인된 폴더들은 캐시에 저장해두고 활용할 수도 있으나... 이 또한 수정이 생기면 반영이 되어야하는데 데이터가 캐시도 온전하지 않을 수 있을 것 같아요
- 게다가 폴더 소유주도 확인해야합니다. (권한 문제)

[확인하지 않는다.]
- 속도는 빠르지만, 폴더가 중간에 끊겨있을 수 있습니다.
- 그런데 그게 큰 문제가 되나? 사실 이것도 판단이 잘 안됩니다.
  - 폴더가 끊겨있을 뿐 파일만 제대로 있다면 다운로드에는 문제가 없는데.. (파일도 다운로드 받을 때 실제로 존재하는지 확인해야겠군요..)
  - 연결이 끊겨있다면 조회에서 어차피 막힐텐데, 그럼 더미값이 남아있긴 한다는 문제가 있고...

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/controller/FileController.java:
```diff
@@ -1,58 +1,69 @@
+        return ResponseEntity.ok()
+            .contentType(MediaType.parseMediaType(file.mime()))
+            .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + file.fileName() + "\"")
+            .body(file.resource());
```
오 좋은데요 배워갑니다

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/service/UserService.java:
```diff
@@ -0,0 +1,60 @@
+	@Transactional
+	public SignUpRes signUp(String userName) {
+		// 유저명 중복 검사
+		checkDuplicate(userName);
```
만약 동시에 유저가 같은 이름으로 회원 가입하면 어떻게 되나요

kochungcheon replied:
동시성 이슈를 처리 할 로직 혹은 락이 필요해보여요

pear96 replied:
https://www.inflearn.com/questions/59250/%EC%95%88%EB%85%95%ED%95%98%EC%84%B8%EC%9A%94-unique-index-%EB%8D%B0%EC%9D%B4%ED%84%B0-%EC%A0%80%EC%9E%A5%EC%97%90-%EB%8C%80%ED%95%B4%EC%84%9C-%EC%A7%88%EB%AC%B8%EB%93%9C%EB%A6%BD%EB%8B%88%EB%8B%A4

이에 따라 unique를 적용하기로 했습니다. (디코 채팅에서 남겨줬었던거 봤어요 땡큐)

pear96 committed:
style: 코드스멜 제거

if문에 블록을 작성하지 않았습니다.

VSFe commented:
> ### 폴더가 제대로 존재하는지(부모 폴더를 타고 올라가서 root까지 닿는지) 확인해야 할까요?
> 폴더가 제대로 존재하는지 확인하려면 최상위까지 타고 올라가봐야하는데, 이걸 한다면 폴더, 파일의 모든 로직마다 수행해야합니다.
> 
> [확인한다.]
> 
> * 요청이 많으면 병목 생기기 딱 좋음. 시간이 매우 많이 소요될 수 있을 것 같습니다..
> * 스스로를 부모 폴더로 보면 안되기 때문에 이제까지 지나온 폴더들이랑 겹치지 않는지도 확인해야합니다. (HashSet..?)
> * 아니면 확인된 폴더들은 캐시에 저장해두고 활용할 수도 있으나... 이 또한 수정이 생기면 반영이 되어야하는데 데이터가 캐시도 온전하지 않을 수 있을 것 같아요
> * 게다가 폴더 소유주도 확인해야합니다. (권한 문제)
> 
> [확인하지 않는다.]
> 
> * 속도는 빠르지만, 폴더가 중간에 끊겨있을 수 있습니다.
> * 그런데 그게 큰 문제가 되나? 사실 이것도 판단이 잘 안됩니다.
>   
>   * 폴더가 끊겨있을 뿐 파일만 제대로 있다면 다운로드에는 문제가 없는데.. (파일도 다운로드 받을 때 실제로 존재하는지 확인해야겠군요..)
>   * 연결이 끊겨있다면 조회에서 어차피 막힐텐데, 그럼 더미값이 남아있긴 한다는 문제가 있고...

여기서는 보정하지 말고 넘어갑시다.

일반적으로는 별도의 정합성 배치를 개발해서 보정하죠?

pear96 committed:
refactor : UUID 이름 메서드 분리

UUID로 서버에 필요한 이름을 만드는 메서드를 분리했습니다.
한줄로 끝나는 코드라 분리해야할 필요가 있을까 싶었지만, 메서드의 이름만으로 용도를 파악하는게
함께 코드를 짜는 환경이라면 가독성도 좋고 이해하기 쉬울것 같아 변경했습니다.

pear96 committed:
refactor : Exception 통합

서비스 예외가 아니라면 공통 예외로 처리하도록 변경했습니다.

pear96 committed:
refactor : 사용자 이름에 Unique 적용

동시성 문제를 해결할 방법을 고민하다가 사용자 이름이 동시에 겹치는 경우는 흔하지 않을 것이라고 생각하여
락을 사용하는 대신 Unique를 적용했습니다.
낙관적 락을 사용하지 않은 이유는 예외 발생시의 롤백, 처리 로직이 필요하기 때문입니다.

pear96 committed:
style : 코드 스멜 제거

VSFe reviewed (approved):
일단 코멘트 더 달고 싶긴 한데... 너무 붙잡아두긴 뭐하니까 스텝 3에서 마저 이야기 합시다.
